src/boot/efi/devicetree.c

```diff
@@ -104,6 +105,91 @@ EFI_STATUS devicetree_install(struct devicetree_state *state, EFI_FILE *root_dir
                         MAKE_GUID_PTR(EFI_DTB_TABLE), PHYSICAL_ADDRESS_TO_POINTER(state->addr));
 }
 
+static char* devicetree_get_compatible(const void *dtb) {
+        assert(dtb);
+        const struct fdt_header *dt_header = dtb;
+        if (!IS_ALIGNED64(dt_header) || be32toh(dt_header->Magic) != UINT32_C(0xd00dfeed))
+                return NULL;
+
+        uint32_t dt_size = be32toh(dt_header->TotalSize);
+        uint32_t struct_off = be32toh(dt_header->OffDTStruct);
+        uint32_t struct_size = be32toh(dt_header->SizeDTStruct);
+        uint32_t strings_off = be32toh(dt_header->OffDTStrings);
+        uint32_t strings_size = be32toh(dt_header->SizeDTStrings);
+
+        if (struct_off % sizeof(uint32_t) || struct_size % sizeof(uint32_t) ||
+            strings_off + strings_size > dt_size || struct_off + struct_size > strings_off)
+                return NULL;
+
+        uint32_t *cursor = (uint32_t *) ((uint8_t *) dt_header + struct_off);
+        char *strings_block = (char *) ((uint8_t *) dt_header + strings_off);
+
+        for (uint32_t i = 0; i < struct_size - 2; i++) {
+                switch (be32toh(cursor[i])) {
+                case FDT_BEGIN_NODE:
+                        if (cursor[++i] != 0)
+                                return NULL;
+                        break;
+                case FDT_NOP:
+                        break;
+                case FDT_PROP:
+                        uint32_t len = be32toh(cursor[++i]);
+                        uint32_t name_off = be32toh(cursor[++i]);
```

Comment:
here too I'd use size_t as types for len and name_off. 

(this has a tangible benefit btw: if you do that then at least on x86-64 the missing overflow check below will be caught)